### PR TITLE
Add num_proc to map and filter calls

### PR DESCRIPTION
**Pull request by dakinggg**
Adds the `num_proc` arg to the calls to huggingface dataset map/filter.

before: `num-proc-test-baseline-1-cDb1Vk`
after: `num-proc-test-1-4nqb1g`

dataset map goes from ~27 seconds to ~4 seconds for dolly_hhrlhf. Should be a bigger win for bigger datasets.

### CODE REVIEW

**Comment by dakinggg** [comment]
@alextrott16 I vaguely recall you had an issue with adding this at one point. Do you remember this? Seems to work fine for me now.

File: llmfoundry/data/finetuning/tasks.py [hunk 1]
```diff
@@ -339,14 +339,19 @@ def dataset_mapper(example: Dict):
...
             return _tokenize_formatted_example(example, tokenizer)
 
+        detected_cpu_count = os.cpu_count() or 1
+
         columns_to_remove = list(dataset[0].keys())
         tokenized_dataset = dataset.map(
             dataset_mapper,
             batched=False,
             remove_columns=columns_to_remove,
+            num_proc=max(1, detected_cpu_count - 4),
```
Comment (irenedea) [inline llmfoundry/data/finetuning/tasks.py:349]: why minus 4?

File: llmfoundry/data/finetuning/tasks.py [hunk 2]
```diff
@@ -339,14 +339,19 @@ def dataset_mapper(example: Dict):
...
         )
         prompt_length_filtered_dataset = tokenized_dataset.filter(
-            lambda example: len(example['input_ids']) < max_seq_len)
+            lambda example: len(example['input_ids']) < max_seq_len,
+            num_proc=max(1, detected_cpu_count - 4),
```
Comment (irenedea) [inline llmfoundry/data/finetuning/tasks.py:353]: super, super nit:  can we use a shared var for L353 and L349

**Review by irenedea — APPROVED** [review]
LGTM! question and super nit

File: llmfoundry/data/finetuning/tasks.py (hunk 1 above)
Comment (dakinggg) [reply llmfoundry/data/finetuning/tasks.py:349]: totally arbitrary choice to not use all the cpus 🤷‍♂️

**Review by alextrott16 — APPROVED** [review]
LGTM. Thanks for doing this.

@dakinggg I don't have a particularly clear memory of running into any issues with using `num_proc`. If you feel confident that our test coverage would catch any issues resulting from this leading to wonkiness, then I'm satisfied.

**Pull request merged**